Add InMemory connector

pnowojski opened the pull request:
Add connector that stores all data in memory on the workers. 

Rationale behind it is to serve as a storage for SQL query benchmarking. Using JMH unit benchmarks from scratch is time consuming to setup, it's often much easier to write some query against TPCH. Previous benchmarks had significant drawback that generating data in TPCH connector was using most of the CPU time. With InMemory connector that's no longer the case.

**Connector is based on BlackHole and first commit is just copy/paste with some renames.**

kokosing commented on pom.xml:
```diff
@@ -81,6 +81,7 @@
         <module>presto-redis</module>
         <module>presto-cassandra</module>
         <module>presto-blackhole</module>
+        <module>presto-in-memory</module>
```
`s/in-memory/inmemory`

Notice that `blackhole` is named `blackhole` not `black-hole`

pnowojski replied:
On the other hand there is `presto-local-file`.  `presto-in-memory` looks better and is more readable (at least for me). Same applies to for example properties: `inmemory.splits-per-node` is worse compared to `in-memory.splits-per-node`.

kokosing commented on presto-in-memory/src/main/java/com/facebook/presto/plugin/inmemory/InMemoryMetadata.java:
```diff
@@ -0,0 +1,193 @@
+    }
+
+    @Override
+    public void finishInsert(ConnectorSession session, ConnectorInsertTableHandle insertHandle, Collection<Slice> fragments) {}
```
it should raise an error that insert is not supported

kokosing commented:
looks good

losipiuk commented on presto-in-memory/src/main/java/com/facebook/presto/plugin/inmemory/InMemoryColumnHandle.java:
```diff
@@ -0,0 +1,104 @@
+import java.util.List;
+import java.util.Objects;
+
+public final class InMemoryColumnHandle
```
Add `toString` method. It is printed in `EXPLAIN` output.

pnowojski replied:
done

losipiuk commented on presto-in-memory/src/main/java/com/facebook/presto/plugin/inmemory/InMemoryTableHandle.java:
```diff
@@ -0,0 +1,112 @@
+
+import static java.util.stream.Collectors.toList;
+
+public final class InMemoryTableHandle
```
add `toString` method

pnowojski replied:
done

losipiuk commented on presto-in-memory/src/main/java/com/facebook/presto/plugin/inmemory/InMemoryInsertTableHandle.java:
```diff
@@ -0,0 +1,37 @@
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+public final class InMemoryInsertTableHandle
```
add `toString` method

pnowojski replied:
done

maciejgrzybek commented on presto-in-memory/src/main/java/com/facebook/presto/plugin/inmemory/InMemoryTableHandle.java:
```diff
@@ -0,0 +1,114 @@
+            @JsonProperty("tableName") String tableName,
+            @JsonProperty("columnHandles") List<InMemoryColumnHandle> columnHandles)
+    {
+        this.connectorId = connectorId;
```
`requireNonNull`

maciejgrzybek commented on presto-in-memory/src/main/java/com/facebook/presto/plugin/inmemory/InMemoryTableLayoutHandle.java:
```diff
@@ -0,0 +1,47 @@
+    @JsonCreator
+    public InMemoryTableLayoutHandle(@JsonProperty("table") InMemoryTableHandle table)
+    {
+        this.table = table;
```
`requireNonNull`

KBP-TDC commented:
@pnowojski should we add a property with a reasonable default to limit how much memory can be used by this connector? An equivalent of "query.max-memory" and "query.max-memory-per-node" I guess.

KBP-TDC commented on presto-in-memory/src/main/java/com/facebook/presto/plugin/inmemory/InMemoryMetadata.java:
```diff
@@ -116,8 +117,9 @@ public ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTable
-        tables.remove(inMemoryTableHandle.getTableName());
+        // TODO: implement this method. Problem is that data are distributed across the cluster and this method
+        // is only being invoked on the coordinator
+        throw new UnsupportedOperationException("DROP TABLE is not supported");
```
Per @maciejgrzybek, we should probably throw `PrestoException` with `NOT_SUPPORTED` `ErrorCode` instead of `java.lang.UnsupportedOperationException`.

maciejgrzybek replied:
Or simply leave default implementation (without overriding) which throws proper exception.

pnowojski replied:
fixed. I kept this method so that I can comment why it's not supported.

maciejgrzybek commented on presto-in-memory/src/test/java/com/facebook/presto/plugin/inmemory/TestInMemorySmoke.java:
```diff
@@ -0,0 +1,141 @@
+            assertTrue(ex.getMessage().equals("line 1:1: Destination table 'inmemory.default.nation' already exists"));
+        }
+        finally {
+            assertThatQueryReturnsValue("DROP TABLE nation", true);
```
If test fails with an exception and `drop` inside `finally` block fails as well, the exception from `finally` will be thrown hiding reason for test failure. You can add cleanup method logging an exception with actual reason of failure. Have a look on `AbstractTestHiveClient` and it's `dropTable` method.

pnowojski replied:
Ok, but it's outdated since `drop table` is not supported, so there is no `finally` block in more recent version.

maciejgrzybek commented on presto-in-memory/src/test/java/com/facebook/presto/plugin/inmemory/TestInMemorySmoke.java:
```diff
@@ -0,0 +1,141 @@
+            assertThatQueryReturnsValue("SELECT count(*) FROM nation", 75L);
+        }
+        finally {
+            assertThatQueryReturnsValue("DROP TABLE nation", true);
```
ditto

pnowojski replied:
as above, comment to outdated diff

maciejgrzybek commented on presto-in-memory/src/main/java/com/facebook/presto/plugin/inmemory/InMemoryInsertTableHandle.java:
```diff
@@ -15,10 +15,23 @@
+    @JsonCreator
+    public InMemoryInsertTableHandle(@JsonProperty("table") InMemoryTableHandle table)
+    {
+        this.table = table;
```
`requireNonNull`

electrum commented:
Let's call this just "Memory" not "InMemory".

sopel39 commented:
Does it distribute data across cluster equally or just stores table data where the writers are? It it stores data on one node (e.g. where the writer op is executed), then there is data skew. Then benchmark results might not be accurate (because of data skew).

pnowojski commented:
@KBP-TDC yes, I was thinking about such property, but as add on since it wasn't required in basic use case (micro benchmarking). Bigger issue is with unsupported `DROP TABLE` which prevents some serious usage of this connector.

pnowojski commented:
@electrum I have renamed it (last commit) but so far everything what was working "in memory" was called `InMemory`, like `InMemoryRecordSet`, `InMemoryJoinHash`. But as you prefer, you can choose whichever version is more appropriate.

KBP-TDC commented:
@pnowojski : Please add documentation listing all limitations and experimental nature of this connector.

pnowojski commented:
@KBP-TDC take look in `memory.rst`. There is clear warning about not supported `DROP TABLE` and no mechanism to release the memory. Do you have something more specific in mind?

KBP-TDC commented on presto-docs/src/main/sphinx/connector/inmemory.rst:
```diff
@@ -0,0 +1,55 @@
+.. warning::
+
+    This connector is in early experimental stage it is not recommended
+    to use it on a production environment.
```
"in a production environment"

pnowojski replied:
done

KBP-TDC commented on presto-docs/src/main/sphinx/connector/inmemory.rst:
```diff
@@ -0,0 +1,55 @@
+
+    When one worker fails/restarts all data that were stored in it's
+    memory will be lost forever. In other words in case of worker
+    restart there will be partial data loss.
```
restart = crash/stop/restart

KBP-TDC replied:
"partial data loss" -> please explain whether users are warned in any way. Or is this silent data loss?

pnowojski replied:
done

KBP-TDC commented on presto-docs/src/main/sphinx/connector/inmemory.rst:
```diff
@@ -0,0 +1,55 @@
+
+.. warning::
+
+    This connector is in early experimental stage it is not recommended
```
Please add something like: "The intended use of this connector at this point is purely for testing"

pnowojski replied:
done

KBP-TDC commented on presto-docs/src/main/sphinx/connector/inmemory.rst:
```diff
@@ -0,0 +1,55 @@
+
+    This connector does not support ``DROP TABLE`` or any other method
+    to release held resources. Once table is populated it will consume
+    memory until next Presto restart.
```
What if only Coordinator is restarted but Workers are not? Is metadata lost but data is still occupying Workers' memory is essentially orphaned?
What happens if Coordinator is not restarted but all Workers are? Table will be "empty" or in invalid state?

pnowojski replied:
updated

KBP-TDC commented on presto-docs/src/main/sphinx/connector/inmemory.rst:
```diff
@@ -0,0 +1,55 @@
+==================
+
+The InMemory connector stores all data and metadata in RAM on workers
+and both are discarded when Presto restarts.
```
"Presto restarts" -> "Presto is restarted on all nodes in the cluster" --- correct?

KBP-TDC replied:
Please add another warning that today there is no way to limit the memory this connector can use. There is also no way to check how much memory is already consumed by the connector on each node.

KBP-TDC commented:
@pnowojski My doc comments are in. Please address them for full disclosure.

pnowojski commented:
After discussion with @kokosing we have figured out some workaround for `DROP TABLE` support. 

We could identify tables by random uuid on the coordinator. `DROP TABLE` would just remove `uuid` from the metadata without doing anything on the workers. To cleanup old tables from the workers, connector would always add to every `ConnectorInsertTableHandle` HashSet with active/up to date/existing uuids. Workers, before writing anything to the tables could scan/validate which uuids that they have in memory are no longer required and can be safely removed. It would work like some lazy garbage collection.

CC @KBP-TDC

electrum commented:
You could use a sequence number instead, and the active set could be a bitset. There are many ways to compress a bitset. A simple way that's useful here given the typical access patten is to include a starting number, so that the leading zeros in the set  can be elided.

pnowojski commented:
@electrum compressing list of active tables doesn't seem like a important feature here. I don't expect more then couple of active/existing tables.

I have added in three last commits support for `DROP TABLE`. Each table has now unique incremental identifier and I'm passing `activeTableIds` set on every write access (on every creation of `PageSink`) which drops unused/old data from the memory.

elonazoulay commented:
Hi, this is a great feature! We pulled in your pull request and made some minor changes (added memory limit and got it working for 156 snapshot) and deployed it to our test cluster:
https://github.com/prestodb/presto/pull/6522

We also wanted to implement stats and help with issues. If there are any new features or changes we would be happy to help contribute.

pnowojski commented:
That's great @elonazoulay, memory limits were on todo list :) Can you point me to the commit that implemented it? I will be happy to add it to this PR.

elonazoulay commented:
Sure, it's: https://github.com/prestodb/presto/pull/6522/commits/484d45e897ad7a6131dd0dba8a8afe97c6a594a4
Let me know if you need any more info or changes.

pnowojski commented:
@elonazoulay I don't see in this commit anything related to 

> added memory limit

Am I missing something?

elonazoulay commented:
@pnowojski Thanks! I added the commit message and updated to work with 0.158-SNAPSHOT: https://github.com/elonazoulay/presto/commit/31aff53c396a3a08290cbc93206442f8c77d4418 -- let me know if there is anything else that needs to be done.

martint commented on presto-benchmark/src/test/java/com/facebook/presto/benchmark/AbstractJmhSqlBenchmark.java:
```diff
@@ -0,0 +1,109 @@
+public class AbstractJmhSqlBenchmark
+{
+    @State(Thread)
+    public static class AbstractContext
```
Why use inheritance? It think it'd be more natural to model this as a state object that the benchmarks can depend on.

pnowojski replied:
fixed

pnowojski commented on presto-memory/src/main/java/com/facebook/presto/plugin/memory/MemoryPagesStore.java:
```diff
@@ -0,0 +1,97 @@
+        // That can lead to a situation when MemoryPagesStore already knows about a newer second table on some worker
+        // but cleanUp is triggered by insert from older first table, which MemoryTableHandle was created before
+        // second table creation. Thus activeTableIds can have missing latest ids and we can only clean up tables
+        // that have smaller value then max(activeTableIds).
```
Case1:

1. insertion into table 100 starts. On coordinator `activeTableIds = {100}`
2. few pages are written on some workers for table `100`. Every worker that has written some pages knows that `activeTableIds = {100}`
3. insertion/creation on table 101 starts. On coordinator `activeTableIds = {100, 101}`
4. if any worker gets some pages from 101 before 100, nothing dramatic happens. No pages are wiped out, because:
  - any `cleanUp` for call with `activeTableIds = {100}` will never delete pages from tables > 100
  - any `cleanUp` for `activeTableIds = {100, 101}` will not drop any pages from 100 nor 101, because they are "active".


After re-reading my comment I know what has mislead you. Sentence: `and we can only clean up tables that have smaller value then max(activeTableIds)`. "Can" doesn't mean it will be delete. Condition for deletion is that table must have:
- `tableId < max(activeTableIds)`
- and table id must be absent in `activeTableIds`

I have modified commit to make that clear.

pnowojski commented on presto-memory/src/main/java/com/facebook/presto/plugin/memory/MemoryConnector.java:
```diff
@@ -0,0 +1,72 @@
+    private final MemoryPageSourceProvider pageSourceProvider;
+    private final MemoryPageSinkProvider pageSinkProvider;
+
+    public MemoryConnector(MemoryMetadata metadata,
```
fixed

pnowojski commented on presto-memory/src/main/java/com/facebook/presto/plugin/memory/MemoryColumnHandle.java:
```diff
@@ -0,0 +1,116 @@
+            @JsonProperty("columnType") Type columnType,
+            @JsonProperty("columnIndex") int columnIndex)
+    {
+        this.name = name;
```
done

pnowojski replied:
ops, fixed

pnowojski commented on presto-memory/src/main/java/com/facebook/presto/plugin/memory/MemoryColumnHandle.java:
```diff
@@ -0,0 +1,116 @@
+            @JsonProperty("columnIndex") int columnIndex)
+    {
+        this.name = name;
+        this.columnType = columnType;
```
done

pnowojski commented on presto-memory/src/main/java/com/facebook/presto/plugin/memory/MemoryPagesStore.java:
```diff
@@ -0,0 +1,97 @@
+@ThreadSafe
+public class MemoryPagesStore
+{
+    private final Map<Long, List<Page>> pages = new HashMap<>();
```
We thought about adding memory tracking as future improvement. Lack of this tracking doesn't impact main use case - automated micro benchmarks and manual macro benchmarks.

pnowojski commented on presto-memory/src/main/java/com/facebook/presto/plugin/memory/MemoryPagesStore.java:
```diff
@@ -0,0 +1,97 @@
+        }
+    }
+
+    private static Page reorderPage(Page page, List<Integer> columnIndexes)
```
done

pnowojski commented on presto-docs/src/main/sphinx/connector/memory.rst:
```diff
@@ -0,0 +1,67 @@
+.. warning::
+
+    After ``DROP TABLE`` memory is not released immediately. It is released
+    after next write access to memory connector.
```
Isn't the `drop table` operation executed on a coordinator only? I don't know if there is some other way to communicate with workers beside `tableHandle` used in `select`, `insert` and `create table`. That's why memory is released on next `insert` or `create` statetment after `drop table`.

haozhun replied:
I should have phrased my question more clearly. I'm specifically asking about why memory is not released on next `select`. I'm not against this. I just want to understand the decision here as this appears like a conscious decision.

pnowojski replied:
It would require both sending `activeTableIds` and add some (small) overhead for checking it and dropping pages on every read access.

On the other hand it doesn't seem to brake anything. Administrator/user must configure this connector and cluster by taking into account maximum potential memory usage by this connector and can't relay that this connector will temporarily use less memory. So it shouldn't matter if the memory is released on first read or first write access.

However I'm not strongly in favor of either version

pnowojski commented on presto-docs/src/main/sphinx/connector/memory.rst:
```diff
@@ -0,0 +1,67 @@
+    memory will be lost forever. In other words in case of worker
+    crash/stop/restart there will be partial and silent data loss.
+    If all worker nodes fail/restart when coordinator is keep
+    running tables will end up empty.
```
In long term we are planning to add those missing features (like memory limits, memory tracking, data loses) to use it reliably on production.

For now it's either for experimental use on production and for benchmarking. We want to restore micro jmh benchmarks using sql queries, similar to old `AbstractBenchmark`, but without the issue of overhead for tpch generation.

pnowojski commented on presto-memory/src/main/java/com/facebook/presto/plugin/memory/MemoryMetadata.java:
```diff
@@ -0,0 +1,217 @@
+    {
+        MemoryTableHandle memoryTableHandle = checkType(tableHandle, MemoryTableHandle.class, "tableHandle");
+        return memoryTableHandle.getColumnHandles().stream()
+                .collect(toMap(MemoryColumnHandle::getName, column -> column));
```
done

pnowojski commented on presto-memory/src/main/java/com/facebook/presto/plugin/memory/MemoryMetadata.java:
```diff
@@ -0,0 +1,217 @@
+    {
+        long nextId = nextTableId.getAndIncrement();
+        Set<Long> activeTableIds = new HashSet(tableIds.values());
+        activeTableIds.add(nextId);
```
indeed.

However there is one bug with transactions. Scenario:

1. beginCreateTable() for table 100
2. beginCreateTable() for table 101
3. inserting some pages to 100 and 101
4. finishCreateTable() for table 101
5. beginCreateTable() for table 102

now any write to table 102 will remove any pages of table 100, because table 100 has not been added to `activeTableIds` yet. I will address this issue later this week or next week.

pnowojski commented on presto-memory/src/main/java/com/facebook/presto/plugin/memory/MemorySplit.java:
```diff
@@ -0,0 +1,121 @@
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+// Right now, splits are just the entire TPCH table
```
fixed

pnowojski commented on presto-memory/src/main/java/com/facebook/presto/plugin/memory/MemorySplit.java:
```diff
@@ -0,0 +1,121 @@
+    private final List<HostAddress> addresses;
+
+    @JsonCreator
+    public MemorySplit(@JsonProperty("tableHandle") MemoryTableHandle tableHandle,
```
fixed

pnowojski commented:
@haozhun there is still one unaddressed @martint's comment. 

Regarding those three big issues:
- Incorrect result when worker dies - please check one of my comments for response
- cleanUp - as above
- memory tracking - with memory limits it will be up to cluster administrator to configure it properly. Besides as in workers dying, for now main purpose is benchmarking. We might want to improve support usage of this connector in the future.

electrum commented:
We can't have a connector that returns wrong answers when a node dies. This is problematic even for testing.

This problem is easy to fix if we have the worker fail the query if they don't have data and we make the splits non-remote (so they can only execute on the correct nodes).

pnowojski commented:
@electrum it was not that simple, but I think I have handled worker dying, restarting and adding new worker with binding list of available workers during table creation, to created table. 

When table is created, all currently available workers are assigned to this table. If new worker is added to the running cluster, it will never participate in reads and writes to the previously created tables. However when worker assigned to a table dies, query reads/writes will fail with: ` No nodes available to run query`. If such worker restarts, query read/writes will fail `Failed to find table on a worker.`

pnowojski commented:
@haozhun kindly reminding about this PR :)

I have rewritten and incorporated @elonazoulay changes that were adding memory limit for stored data per node.

pnowojski commented:
@martint, @electrum is there any pending issue with this connector?

haozhun commented:
@pnowojski Sorry about the delay. I'll be back in the office 2nd week of January. I'll review this pull request then.

haozhun commented on presto-memory/src/main/java/com/facebook/presto/plugin/memory/MemoryMetadata.java:
```diff
@@ -0,0 +1,231 @@
+        tableIds.remove(oldTableHandle.getTableName());
+        tableIds.put(newTableName.getTableName(), oldTableHandle.getTableId());
+        tables.remove(oldTableHandle.getTableId());
+        tables.put(oldTableHandle.getTableId(), newTableHandle);
```
Thread safety for `tables` and `tableIds` are concerning.

* Sequence of `tables.remove(x); tables.put(x, ...);` can result in the table being pruned in the middle.
* `ImmutableSet.copyOf(tableIds.values)` may observe mixed state. It is not straightforward to reason about whether such observation is guaranteed to be acceptable or not.

Back to this comment about unintended deletion I left during my first-pass: https://github.com/prestodb/presto/pull/5809#discussion_r88376531

* I came up with those examples under the impression that update to `tableIds` and `tables` are carried out during `finishCreateTable` and `finishInsert`. I can see that this is not the case now. It looks like my memory didn't serve me well when I made those examples.
* The general problem still exists with the current code however. In `beginCreateTable`, increment to `nextTableId` and insertion to `tableIds` can interleave. As a result, smaller table ids are not guaranteed to be inserted into the maps prior to the larger table ids. This can lead to unintended deletion of table data (i.e. with some tweaks, the scenarios I described in the previous comment can happen because of this bullet point.)

It seems like it would just be so much easier to reason about these things if we protect all these three variables with a synchronized block. It doesn't look like concurrency performance is a concern for this class.

pnowojski replied:
Hmmm, good catch. I assumed that `ConnectorMetadata` is used only by single thread.

I will synchronize all methods in `MemoryMetadata`. I think that should solve those problems and indeed synchronization is not an issue here (at least now).

pnowojski commented on presto-main/etc/config.properties:
```diff
@@ -36,7 +37,7 @@ plugin.bundles=\
   ../presto-local-file/pom.xml, \
   ../presto-mysql/pom.xml,\
-  ../presto-postgresql/pom.xml
+  ../presto-postgresql/pom.xml,\
```
reverted

pnowojski commented on presto-main/etc/catalog/memory.properties:
```diff
@@ -0,0 +1 @@
+connector.name=memory
```
fixed

pnowojski commented on presto-benchmark/src/test/java/com/facebook/presto/benchmark/MemoryLocalQueryRunner.java:
```diff
@@ -0,0 +1,89 @@
+{
+    protected LocalQueryRunner localQueryRunner = createMemoryLocalQueryRunner();
+
+    public void execute(@Language("SQL") String query)
```
`LocaleQueryRunner::execute` materializes output, which can skew benchmark results. This one, uses `NullOutputFactory`.

pnowojski commented on presto-benchmark/pom.xml:
```diff
@@ -92,6 +97,18 @@
 
         <!-- for testing -->
         <dependency>
+            <groupId>org.openjdk.jmh</groupId>
```
move to main

pnowojski replied:
I have moved it to main, and later then I have remembered why I placed it in test. Jmh and dependency checker plugin doesn play well with:

```
        <dependency>
            <groupId>org.openjdk.jmh</groupId>
            <artifactId>jmh-generator-annprocess</artifactId>
        </dependency>
```

This is required to run benchmarks, but dependency checker fails with "unused declared dependency". If it's marked as `<scope>runtime</scope>` or `<scope>test</scope>`, jmh then fails:

```
/usr/lib/jvm/java-8-oracle/bin/java -Dfile.encoding=UTF-8 -classpath /home/piotr/presto-td/presto-benchmark/target/classes:/home/piotr/presto-td/presto-spi/target/classes:/home/piotr/.m2/repository/io/airlift/slice/0.27/slice-0.27.jar:/home/piotr/.m2/repository/org/openjdk/jol/jol-core/0.2/jol-core-0.2.jar:/home/piotr/presto-td/presto-main/target/classes:/home/piotr/.m2/repository/it/unimi/dsi/fastutil/6.5.9/fastutil-6.5.9.jar:/home/piotr/.m2/repository/org/apache/commons/commons-math3/3.2/commons-math3-3.2.jar:/home/piotr/presto-td/presto-array/target/classes:/home/piotr/presto-td/presto-client/target/classes:/home/piotr/presto-td/presto-parser/target/classes:/home/piotr/.m2/repository/org/antlr/antlr4-runtime/4.6/antlr4-runtime-4.6.jar:/home/piotr/presto-td/presto-bytecode/target/classes:/home/piotr/.m2/repository/io/airlift/concurrent/0.139/concurrent-0.139.jar:/home/piotr/.m2/repository/io/airlift/node/0.139/node-0.139.jar:/home/piotr/.m2/repository/io/airlift/configuration/0.139/configuration-0.139.jar:/home/piotr/.m2/repository/cglib/cglib-nodep/2.2.2/cglib-nodep-2.2.2.jar:/home/piotr/.m2/repository/io/airlift/discovery/0.139/discovery-0.139.jar:/home/piotr/.m2/repository/io/airlift/event/0.139/event-0.139.jar:/home/piotr/.m2/repository/io/airlift/http-server/0.139/http-server-0.139.jar:/home/piotr/.m2/repository/org/eclipse/jetty/jetty-http/9.3.13.M0/jetty-http-9.3.13.M0.jar:/home/piotr/.m2/repository/org/eclipse/jetty/jetty-io/9.3.13.M0/jetty-io-9.3.13.M0.jar:/home/piotr/.m2/repository/org/eclipse/jetty/http2/http2-server/9.3.13.M0/http2-server-9.3.13.M0.jar:/home/piotr/.m2/repository/org/eclipse/jetty/http2/http2-common/9.3.13.M0/http2-common-9.3.13.M0.jar:/home/piotr/.m2/repository/org/eclipse/jetty/http2/http2-hpack/9.3.13.M0/http2-hpack-9.3.13.M0.jar:/home/piotr/.m2/repository/org/eclipse/jetty/jetty-server/9.3.13.M0/jetty-server-9.3.13.M0.jar:/home/piotr/.m2/repository/org/eclipse/jetty/jetty-util/9.3.13.M0/jetty-util-9.3.13.M0.jar:/home/piotr/.m2/repository/org/eclipse/jetty/jetty-servlet/9.3.13.M0/jetty-servlet-9.3.13.M0.jar:/home/piotr/.m2/repository/org/eclipse/jetty/jetty-security/9.3.13.M0/jetty-security-9.3.13.M0.jar:/home/piotr/.m2/repository/org/eclipse/jetty/jetty-jmx/9.3.13.M0/jetty-jmx-9.3.13.M0.jar:/home/piotr/.m2/repository/ch/qos/logback/logback-core/1.0.13/logback-core-1.0.13.jar:/home/piotr/.m2/repository/io/airlift/jaxrs/0.139/jaxrs-0.139.jar:/home/piotr/.m2/repository/javax/activation/activation/1.1.1/activation-1.1.1.jar:/home/piotr/.m2/repository/org/glassfish/jersey/core/jersey-common/2.22.2/jersey-common-2.22.2.jar:/home/piotr/.m2/repository/org/glassfish/jersey/bundles/repackaged/jersey-guava/2.22.2/jersey-guava-2.22.2.jar:/home/piotr/.m2/repository/org/glassfish/hk2/hk2-api/2.4.0-b34/hk2-api-2.4.0-b34.jar:/home/piotr/.m2/repository/org/glassfish/hk2/hk2-utils/2.4.0-b34/hk2-utils-2.4.0-b34.jar:/home/piotr/.m2/repository/org/glassfish/hk2/hk2-locator/2.4.0-b34/hk2-locator-2.4.0-b34.jar:/home/piotr/.m2/repository/org/javassist/javassist/3.18.1-GA/javassist-3.18.1-GA.jar:/home/piotr/.m2/repository/org/glassfish/hk2/osgi-resource-locator/1.0.1/osgi-resource-locator-1.0.1.jar:/home/piotr/.m2/repository/org/glassfish/jersey/core/jersey-server/2.22.2/jersey-server-2.22.2.jar:/home/piotr/.m2/repository/org/glassfish/jersey/core/jersey-client/2.22.2/jersey-client-2.22.2.jar:/home/piotr/.m2/repository/org/glassfish/jersey/media/jersey-media-jaxb/2.22.2/jersey-media-jaxb-2.22.2.jar:/home/piotr/.m2/repository/org/glassfish/jersey/containers/jersey-container-servlet-core/2.22.2/jersey-container-servlet-core-2.22.2.jar:/home/piotr/.m2/repository/org/glassfish/jersey/containers/jersey-container-servlet/2.22.2/jersey-container-servlet-2.22.2.jar:/home/piotr/.m2/repository/com/fasterxml/jackson/dataformat/jackson-dataformat-smile/2.8.1/jackson-dataformat-smile-2.8.1.jar:/home/piotr/.m2/repository/io/airlift/jmx/0.139/jmx-0.139.jar:/home/piotr/.m2/repository/io/airlift/jmx-http/0.139/jmx-http-0.139.jar:/home/piotr/.m2/repository/io/airlift/log-manager/0.139/log-manager-0.139.jar:/home/piotr/.m2/repository/org/slf4j/slf4j-jdk14/1.7.12/slf4j-jdk14-1.7.12.jar:/home/piotr/.m2/repository/org/slf4j/log4j-over-slf4j/1.7.12/log4j-over-slf4j-1.7.12.jar:/home/piotr/.m2/repository/org/slf4j/jcl-over-slf4j/1.7.12/jcl-over-slf4j-1.7.12.jar:/home/piotr/.m2/repository/io/airlift/resolver/resolver/1.3/resolver-1.3.jar:/home/piotr/.m2/repository/org/sonatype/aether/aether-spi/1.13.1/aether-spi-1.13.1.jar:/home/piotr/.m2/repository/org/sonatype/aether/aether-impl/1.13.1/aether-impl-1.13.1.jar:/home/piotr/.m2/repository/org/sonatype/aether/aether-util/1.13.1/aether-util-1.13.1.jar:/home/piotr/.m2/repository/org/sonatype/aether/aether-connector-file/1.13.1/aether-connector-file-1.13.1.jar:/home/piotr/.m2/repository/org/sonatype/aether/aether-connector-asynchttpclient/1.13.1/aether-connector-asynchttpclient-1.13.1.jar:/home/piotr/.m2/repository/com/ning/async-http-client/1.6.5/async-http-client-1.6.5.jar:/home/piotr/.m2/repository/io/netty/netty/3.7.0.Final/netty-3.7.0.Final.jar:/home/piotr/.m2/repository/org/apache/maven/maven-core/3.0.4/maven-core-3.0.4.jar:/home/piotr/.m2/repository/org/apache/maven/maven-settings/3.0.4/maven-settings-3.0.4.jar:/home/piotr/.m2/repository/org/apache/maven/maven-settings-builder/3.0.4/maven-settings-builder-3.0.4.jar:/home/piotr/.m2/repository/org/apache/maven/maven-repository-metadata/3.0.4/maven-repository-metadata-3.0.4.jar:/home/piotr/.m2/repository/org/apache/maven/maven-plugin-api/3.0.4/maven-plugin-api-3.0.4.jar:/home/piotr/.m2/repository/org/apache/maven/maven-model-builder/3.0.4/maven-model-builder-3.0.4.jar:/home/piotr/.m2/repository/org/codehaus/plexus/plexus-interpolation/1.14/plexus-interpolation-1.14.jar:/home/piotr/.m2/repository/org/codehaus/plexus/plexus-utils/2.0.6/plexus-utils-2.0.6.jar:/home/piotr/.m2/repository/org/codehaus/plexus/plexus-component-annotations/1.5.5/plexus-component-annotations-1.5.5.jar:/home/piotr/.m2/repository/org/sonatype/plexus/plexus-sec-dispatcher/1.3/plexus-sec-dispatcher-1.3.jar:/home/piotr/.m2/repository/org/apache/maven/maven-model/3.0.4/maven-model-3.0.4.jar:/home/piotr/.m2/repository/org/apache/maven/maven-artifact/3.0.4/maven-artifact-3.0.4.jar:/home/piotr/.m2/repository/org/apache/maven/maven-aether-provider/3.0.4/maven-aether-provider-3.0.4.jar:/home/piotr/.m2/repository/org/apache/maven/maven-embedder/3.0.4/maven-embedder-3.0.4.jar:/home/piotr/.m2/repository/org/apache/maven/maven-compat/3.0.4/maven-compat-3.0.4.jar:/home/piotr/.m2/repository/org/apache/maven/wagon/wagon-provider-api/2.2/wagon-provider-api-2.2.jar:/home/piotr/.m2/repository/org/sonatype/plexus/plexus-cipher/1.7/plexus-cipher-1.7.jar:/home/piotr/.m2/repository/org/codehaus/plexus/plexus-container-default/1.5.5/plexus-container-default-1.5.5.jar:/home/piotr/.m2/repository/org/apache/xbean/xbean-reflect/3.4/xbean-reflect-3.4.jar:/home/piotr/.m2/repository/org/codehaus/plexus/plexus-classworlds/2.4/plexus-classworlds-2.4.jar:/home/piotr/.m2/repository/org/slf4j/slf4j-api/1.7.12/slf4j-api-1.7.12.jar:/home/piotr/.m2/repository/io/airlift/trace-token/0.139/trace-token-0.139.jar:/home/piotr/.m2/repository/io/airlift/joni/2.1.5.1/joni-2.1.5.1.jar:/home/piotr/.m2/repository/com/teradata/re2j-td/1.4/re2j-td-1.4.jar:/home/piotr/.m2/repository/io/airlift/discovery/discovery-server/1.27/discovery-server-1.27.jar:/home/piotr/.m2/repository/io/airlift/jmx-http-rpc/0.136/jmx-http-rpc-0.136.jar:/home/piotr/.m2/repository/org/iq80/leveldb/leveldb-api/0.8/leveldb-api-0.8.jar:/home/piotr/.m2/repository/org/iq80/leveldb/leveldb/0.8/leveldb-0.8.jar:/home/piotr/.m2/repository/javax/servlet/javax.servlet-api/3.1.0/javax.servlet-api-3.1.0.jar:/home/piotr/.m2/repository/javax/annotation/javax.annotation-api/1.2/javax.annotation-api-1.2.jar:/home/piotr/.m2/repository/com/fasterxml/jackson/core/jackson-databind/2.8.1/jackson-databind-2.8.1.jar:/home/piotr/.m2/repository/joda-time/joda-time/2.8.2/joda-time-2.8.2.jar:/home/piotr/.m2/repository/org/weakref/jmxutils/1.19/jmxutils-1.19.jar:/home/piotr/.m2/repository/javax/validation/validation-api/1.1.0.Final/validation-api-1.1.0.Final.jar:/home/piotr/.m2/repository/javax/inject/javax.inject/1/javax.inject-1.jar:/home/piotr/.m2/repository/com/google/inject/guice/4.0/guice-4.0.jar:/home/piotr/.m2/repository/aopalliance/aopalliance/1.0/aopalliance-1.0.jar:/home/piotr/.m2/repository/com/google/inject/extensions/guice-multibindings/4.0/guice-multibindings-4.0.jar:/home/piotr/.m2/repository/io/airlift/http-client/0.139/http-client-0.139.jar:/home/piotr/.m2/repository/org/eclipse/jetty/jetty-client/9.3.13.M0/jetty-client-9.3.13.M0.jar:/home/piotr/.m2/repository/org/eclipse/jetty/http2/http2-client/9.3.13.M0/http2-client-9.3.13.M0.jar:/home/piotr/.m2/repository/org/eclipse/jetty/jetty-alpn-client/9.3.13.M0/jetty-alpn-client-9.3.13.M0.jar:/home/piotr/.m2/repository/org/eclipse/jetty/http2/http2-http-client-transport/9.3.13.M0/http2-http-client-transport-9.3.13.M0.jar:/home/piotr/.m2/repository/org/sonatype/aether/aether-api/1.13.1/aether-api-1.13.1.jar:/home/piotr/.m2/repository/org/ow2/asm/asm-all/5.0.4/asm-all-5.0.4.jar:/home/piotr/.m2/repository/org/jgrapht/jgrapht-core/0.9.0/jgrapht-core-0.9.0.jar:/home/piotr/.m2/repository/org/apache/bval/bval-jsr/1.1.1/bval-jsr-1.1.1.jar:/home/piotr/.m2/repository/org/apache/bval/bval-core/1.1.1/bval-core-1.1.1.jar:/home/piotr/.m2/repository/commons-beanutils/commons-beanutils-core/1.8.3/commons-beanutils-core-1.8.3.jar:/home/piotr/.m2/repository/org/apache/commons/commons-lang3/3.4/commons-lang3-3.4.jar:/home/piotr/.m2/repository/javax/xml/bind/jaxb-api/2.2.6/jaxb-api-2.2.6.jar:/home/piotr/.m2/repository/com/sun/xml/bind/jaxb-impl/2.2.6/jaxb-impl-2.2.6.jar:/home/piotr/presto-td/presto-tpch/target/classes:/home/piotr/.m2/repository/io/airlift/tpch/tpch/0.8/tpch-0.8.jar:/home/piotr/presto-td/presto-memory/target/classes:/home/piotr/.m2/repository/com/google/code/findbugs/annotations/2.0.3/annotations-2.0.3.jar:/home/piotr/.m2/repository/io/airlift/json/0.139/json-0.139.jar:/home/piotr/.m2/repository/com/fasterxml/jackson/datatype/jackson-datatype-jdk8/2.8.1/jackson-datatype-jdk8-2.8.1.jar:/home/piotr/.m2/repository/com/fasterxml/jackson/datatype/jackson-datatype-jsr310/2.8.1/jackson-datatype-jsr310-2.8.1.jar:/home/piotr/.m2/repository/com/fasterxml/jackson/datatype/jackson-datatype-guava/2.8.1/jackson-datatype-guava-2.8.1.jar:/home/piotr/.m2/repository/com/fasterxml/jackson/datatype/jackson-datatype-joda/2.8.1/jackson-datatype-joda-2.8.1.jar:/home/piotr/.m2/repository/io/airlift/log/0.139/log-0.139.jar:/home/piotr/.m2/repository/io/airlift/units/1.0/units-1.0.jar:/home/piotr/.m2/repository/com/fasterxml/jackson/core/jackson-annotations/2.8.1/jackson-annotations-2.8.1.jar:/home/piotr/.m2/repository/com/fasterxml/jackson/core/jackson-core/2.8.1/jackson-core-2.8.1.jar:/home/piotr/.m2/repository/com/google/guava/guava/18.0/guava-18.0.jar:/home/piotr/.m2/repository/org/openjdk/jmh/jmh-core/1.15/jmh-core-1.15.jar:/home/piotr/.m2/repository/net/sf/jopt-simple/jopt-simple/4.6/jopt-simple-4.6.jar:/home/piotr/.m2/repository/io/airlift/stats/0.139/stats-0.139.jar:/home/piotr/.m2/repository/org/hdrhistogram/HdrHistogram/2.1.9/HdrHistogram-2.1.9.jar:/home/piotr/.m2/repository/io/airlift/bootstrap/0.139/bootstrap-0.139.jar:/home/piotr/.m2/repository/javax/ws/rs/javax.ws.rs-api/2.0.1/javax.ws.rs-api-2.0.1.jar:/home/piotr/.m2/repository/org/openjdk/jmh/jmh-generator-annprocess/1.15/jmh-generator-annprocess-1.15.jar org.openjdk.jmh.Main com.facebook.presto.benchmark.BenchmarkDecimalAggregation.benchmarkBuildHash
Exception in thread "main" java.lang.RuntimeException: ERROR: Unable to find the resource: /META-INF/BenchmarkList
	at org.openjdk.jmh.runner.AbstractResourceReader.getReaders(AbstractResourceReader.java:97)
	at org.openjdk.jmh.runner.BenchmarkList.find(BenchmarkList.java:103)
	at org.openjdk.jmh.runner.Runner.internalRun(Runner.java:256)
	at org.openjdk.jmh.runner.Runner.run(Runner.java:206)
	at org.openjdk.jmh.Main.main(Main.java:71)
```

I'm leaving it in test for now.

pnowojski commented on presto-benchmark/src/test/java/com/facebook/presto/benchmark/BenchmarkDecimalAggregation.java:
```diff
@@ -0,0 +1,72 @@
+    public void benchmarkBuildHash(AggregationContext context)
+    {
+        context.run();
+    }
```
? I'm using JMH plugin and I have no problems at all with that.

pnowojski replied:
regardless of the answer, I have added `main` method.

pnowojski commented on presto-memory/src/main/java/com/facebook/presto/plugin/memory/MemoryColumnHandle.java:
```diff
@@ -0,0 +1,117 @@
+    {
+        this.name = name;
+        this.columnType = requireNonNull(columnType, "columnType is null");
+        this.columnIndex = requireNonNull(columnIndex, "columnIndex is null");
```
fixed

pnowojski commented on presto-memory/src/main/java/com/facebook/presto/plugin/memory/MemoryConnectorFactory.java:
```diff
@@ -0,0 +1,88 @@
+    public Connector create(String connectorId, Map<String, String> requiredConfig, ConnectorContext context)
+    {
+        int splitsPerNode = getSplitsPerNode(requiredConfig);
+        DataSize maxDataPerNode = getMaxDataPerNode(requiredConfig);
```
Ok, I have moved it to guice and used config class.

pnowojski commented:
Thanks for the review @haozhun :) I haven't managed to review and apply all your comments, I will finish it tomorrow.

pnowojski commented on presto-memory/src/main/java/com/facebook/presto/plugin/memory/MemoryMetadata.java:
```diff
@@ -0,0 +1,231 @@
+        if (tableId == null) {
+            return null;
+        }
+        return tables.get(tableId).updateActiveTableIds(ImmutableSet.copyOf(tableIds.values()));
```
I have moved `activeTableIds` from `MemoryTableHandle` to `MemoryInsertTableHandle` and `MemoryOutputTableHandle`

pnowojski commented on presto-memory/src/main/java/com/facebook/presto/plugin/memory/MemoryMetadata.java:
```diff
@@ -0,0 +1,231 @@
+                nextId,
+                tableMetadata,
+                nodes.stream().map(Node::getHostAndPort).collect(Collectors.toList()),
+                ImmutableSet.copyOf(tableIds.values()));
```
I have moved `activeTableIds` to output and insert table handles.

pnowojski commented on presto-memory/src/main/java/com/facebook/presto/plugin/memory/MemoryPageSourceProvider.java:
```diff
@@ -0,0 +1,61 @@
+        MemorySplit memorySplit = checkType(split, MemorySplit.class, "split");
+        long tableId = memorySplit.getTableHandle().getTableId();
+        int partNumber = memorySplit.getPartNumber();
+        int totalParts = memorySplit.getTotalPartsPerWorker();
```
I have added some comments

pnowojski commented on presto-memory/src/main/java/com/facebook/presto/plugin/memory/MemoryPagesStore.java:
```diff
@@ -0,0 +1,130 @@
+            return;
+        }
+        long latestTableId  = Collections.max(activeTableIds);
+        for (Long tableId : ImmutableList.copyOf(pages.keySet())) {
```
As you wish, however standard practice is utterly ugly and unreadable because of all that boilerplate:

```
        for (Iterator<Map.Entry<Long, List<Page>>> tablePages = pages.entrySet().iterator(); tablePages.hasNext();) {
            Map.Entry<Long, List<Page>> tablePagesEntry = tablePages.next();
            Long tableId = tablePagesEntry.getKey();
            if (tableId < latestTableId && !activeTableIds.contains(tableId)) {
                for (Page removedPage : tablePagesEntry.getValue()) {
                    currentBytes -= removedPage.getRetainedSizeInBytes();
                }
                tablePages.remove();
            }
        }
```

pnowojski commented on presto-memory/src/main/java/com/facebook/presto/plugin/memory/MemorySplitManager.java:
```diff
@@ -0,0 +1,65 @@
+
+        List<HostAddress> hosts = layout.getTable().getHosts();
+
+        // Split the data using split and skew by the number of nodes available.
```
Hmm, I have dropped this comment, because this code seems self explanatory and there are comments in `MemorySplits` explaining its fields.

pnowojski commented on presto-memory/src/main/java/com/facebook/presto/plugin/memory/Types.java:
```diff
@@ -0,0 +1,33 @@
+{
+    private Types() {}
+
+    public static <A, B extends A> B checkType(A value, Class<B> target, String name)
```
I have removed this in separate commit. It would be tiresome to found in which commit every `checkType` usage was introduced and fix it that way :/

pnowojski commented:
@haozhun I have applied all your comments except of moving benchmarks to src/main (please check my response there)

pnowojski commented:
@haozhun ?

pnowojski commented:
@haozhun @martint  please merge this, since more and more PRs are depending on benchmark framework defined in this PR. We are waiting half year for those benchmarks and during this time, this connector evolved to something much bigger then it was intended (95% of comments that I have applied had nothing to do with jmh benchmarking using this connector). 

If there are some more missing features that prevents this connector from usage on production, that is completely fine with me. Those issues can be resolved in future PRs, while we need THIS PR for easy to write benchmarks.

haozhun commented:
Looks good. I'll squash, merge, and update versions in pom once release is out.

pnowojski commented:
Thanks a lot.

elonazoulay commented:
Hi, we have been using this connector in our deployment and found one minor issue:

Line 32 in MemoryConfig.java has "metadata-uri" for the config variable name for setSplitsPerWorker, should it be "splits-per-worker" or something like that?

BTW, this connector is an awesome feature!

pnowojski commented:
Yes @elonazoulay , definitely it should be `splits-per-node`. I have fixed it and pushed small fixup.

haozhun commented:
Merged #5809

pnowojski commented:
Thanks :)
